sql/sem/builtins: TestGetSSTableMetricsSingleNode failed

sql/sem/builtins.TestGetSSTableMetricsSingleNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20330586?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20330586?buildTab=artifacts#/) on master @ [bb8c1ebbc4074a54f322b2fdfd4505a43d099f89](https://github.com/cockroachdb/cockroach/commits/bb8c1ebbc4074a54f322b2fdfd4505a43d099f89):


```
=== RUN   TestGetSSTableMetricsSingleNode
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/9ec469362276cb97645f959c139369c8/logTestGetSSTableMetricsSingleNode3405000609
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:143: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    generator_builtins_test.go:198: 
        	Error Trace:	pkg/sql/sem/builtins/generator_builtins_test.go:198
        	Error:      	"0" is not greater than or equal to "1"
        	Test:       	TestGetSSTableMetricsSingleNode
    panic.go:636: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9ec469362276cb97645f959c139369c8/logTestGetSSTableMetricsSingleNode3405000609
--- FAIL: TestGetSSTableMetricsSingleNode (4.69s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGetSSTableMetricsSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53452

sql/sem/builtins.TestGetSSTableMetricsSingleNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20421108?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20421108?buildTab=artifacts#/) on master @ [9a77ab62090722887cc70dc1d33c1345a2d675a6](https://github.com/cockroachdb/cockroach/commits/9a77ab62090722887cc70dc1d33c1345a2d675a6):


```
=== RUN   TestGetSSTableMetricsSingleNode
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/9ec469362276cb97645f959c139369c8/logTestGetSSTableMetricsSingleNode1875859228
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:143: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    generator_builtins_test.go:198: 
        	Error Trace:	pkg/sql/sem/builtins/generator_builtins_test.go:198
        	Error:      	"0" is not greater than or equal to "1"
        	Test:       	TestGetSSTableMetricsSingleNode
    panic.go:635: -- test log scope end --
--- FAIL: TestGetSSTableMetricsSingleNode (4.84s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGetSSTableMetricsSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/sem/builtins.TestGetSSTableMetricsSingleNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20501039?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20501039?buildTab=artifacts#/) on master @ [4dd6f9e3ceeecbf5277b4baee2336fcfd86a17fc](https://github.com/cockroachdb/cockroach/commits/4dd6f9e3ceeecbf5277b4baee2336fcfd86a17fc):


```
=== RUN   TestGetSSTableMetricsSingleNode
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/9ec469362276cb97645f959c139369c8/logTestGetSSTableMetricsSingleNode3848140082
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:141: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    generator_builtins_test.go:198: 
        	Error Trace:	pkg/sql/sem/builtins/generator_builtins_test.go:198
        	Error:      	"0" is not greater than or equal to "1"
        	Test:       	TestGetSSTableMetricsSingleNode
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/9ec469362276cb97645f959c139369c8/logTestGetSSTableMetricsSingleNode3848140082
--- FAIL: TestGetSSTableMetricsSingleNode (2.39s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGetSSTableMetricsSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/sem/builtins/generator_builtins_test.go
+++ b/pkg/sql/sem/builtins/generator_builtins_test.go
@@ -11,12 +11,14 @@ import (
 	"encoding/json"
 	"fmt"
 	"math/rand"
+	"slices"
 	"testing"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
@@ -157,27 +159,30 @@ func TestGetSSTableMetricsSingleNode(t *testing.T) {
 	ts, hostDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
 	defer ts.Stopper().Stop(ctx)
 
-	nodeIDArg := 1
+	nodeIDArg := int(ts.NodeID())
 	storeIDArg := int(ts.GetFirstStoreID())
 
 	r := sqlutils.MakeSQLRunner(hostDB)
-	r.Exec(t, `CREATE TABLE t(k INT PRIMARY KEY, v INT)`)
-	r.Exec(t, `INSERT INTO t SELECT i, i*10 FROM generate_series(1, 10000) AS g(i)`)
 
-	r.Exec(t, fmt.Sprintf(`
-	 SELECT crdb_internal.compact_engine_span(
-		 %d, %d,
-		 (SELECT raw_start_key FROM [SHOW RANGES FROM TABLE t WITH KEYS] LIMIT 1),
-		 (SELECT raw_end_key FROM [SHOW RANGES FROM TABLE t WITH KEYS] LIMIT 1))`,
-		nodeIDArg, storeIDArg))
+	// Seed some data.
+	r.Exec(t, `CREATE TABLE t(k INT PRIMARY KEY, v STRING)`)
+	r.Exec(t, `INSERT INTO t SELECT i, CAST(gen_random_uuid() AS STRING) FROM generate_series(1, 10000) AS g(i)`)
 
-	rows := r.Query(t, fmt.Sprintf(`
-	 SELECT * FROM crdb_internal.sstable_metrics(
-		 %d, %d,
-		 (SELECT raw_start_key FROM [SHOW RANGES FROM TABLE t WITH KEYS] LIMIT 1),
-		 (SELECT raw_end_key FROM [SHOW RANGES FROM TABLE t WITH KEYS] LIMIT 1))`,
-		nodeIDArg, storeIDArg))
+	// Construct min and max engine keys. We append a 0x00 byte to the end of
+	// each to make them valid engine keys (the last byte indicates the length
+	// of the version, in this case none).
+	minEngineKey := append(slices.Clone(roachpb.KeyMin), 0x00)
+	maxEngineKey := append(slices.Clone(roachpb.KeyMax), 0x00)
 
+	// Manually compact the entire user key space. This will trigger a memtable
+	// flush if needed, ensuring that the data written above has been flushed to
+	// sstables.
+	r.Exec(t, `SELECT crdb_internal.compact_engine_span($1, $2, $3, $4)`,
+		nodeIDArg, storeIDArg, minEngineKey, maxEngineKey)
+
+	// Now there must exist at least one sstable.
+	rows := r.Query(t, `SELECT * FROM crdb_internal.sstable_metrics($1, $2, $3, $4)`,
+		nodeIDArg, storeIDArg, minEngineKey, maxEngineKey)
 	count := 0
 	var nodeID int
 	var storeID int
@@ -195,6 +200,7 @@ func TestGetSSTableMetricsSingleNode(t *testing.T) {
 		require.NotEqual(t, approximateSpanBytes, 0)
 		count++
 	}
+	require.NoError(t, rows.Err())
 	require.GreaterOrEqual(t, count, 1)
 }
 

